packages/core/src/notifications/controller.ts

```diff
@@ -0,0 +1,233 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { ToolkitError } from '../shared/errors'
+import globals from '../shared/extensionGlobals'
+import { globalKey } from '../shared/globalState'
+import { NotificationsState, NotificationType, NotificationData, ToolkitNotification } from './types'
+import { HttpResourceFetcher } from '../shared/resourcefetcher/httpResourceFetcher'
+import { getLogger } from '../shared/logger/logger'
+import { NotificationsNode } from './panelNode'
+import { Commands } from '../shared/vscode/commands2'
+import { RuleEngine } from './rules'
+import { TreeNode } from '../shared/treeview/resourceTreeDataProvider'
+import { withRetries } from '../shared/utilities/functionUtils'
+import { FileResourceFetcher } from '../shared/resourcefetcher/fileResourceFetcher'
+
+const startUpEndpoint = 'https://idetoolkits-hostedfiles.amazonaws.com/Notifications/VSCode/startup/1.x.json'
+const emergencyEndpoint = 'https://idetoolkits-hostedfiles.amazonaws.com/Notifications/VSCode/emergency/1.x.json'
+
+type ResourceResponse = Awaited<ReturnType<HttpResourceFetcher['getNewETagContent']>>
+
+/**
+ * Handles fetching and maintaining the state of in-IDE notifications.
+ * Notifications are constantly polled from a known endpoint and then stored in global state.
+ * The global state is used to compare if there are a change in notifications on the endpoint
+ * or if the endpoint is not reachable.
+ *
+ * This class will send any notifications to {@link NotificationsNode} for display.
+ * Notifications can be dismissed.
+ *
+ * Startup notifications - fetched each start up.
+ * Emergency notifications - fetched at a regular interval.
+ */
+export class NotificationsController {
+    public static readonly retryNumber = 5
+    public static readonly retryIntervalMs = 30000
+    public static readonly suggestedPollIntervalMs = 1000 * 60 * 10 // 10 minutes
+
+    public readonly storageKey: globalKey
+
+    /** Internal memory state that is written to global state upon modification. */
+    private readonly state: NotificationsState
+    private readonly notificationsNode: NotificationsNode
+
+    static #instance: NotificationsController | undefined
+
+    constructor(extPrefix: 'amazonq' | 'toolkit', node: NotificationsNode) {
+        if (!NotificationsController.#instance) {
+            registerDismissCommand(extPrefix)
+        }
+        NotificationsController.#instance = this
+
+        this.storageKey = `aws.${extPrefix}.notifications`
+        this.notificationsNode = node
+
+        this.state = globals.globalState.get(this.storageKey) ?? {
```

Comment:
generally a good idea to use `tryGet`. it gives some extra type safety.
```suggestion
        this.state = globals.globalState.tryGet(this.storageKey, Object, {
```